File: 22-generate-parentheses/generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        ans = []

        def solve(temp, cntx, cnty):
            if cntx > cnty or cntx < 0:
                return
            if cntx==0 and cnty==0:
                ans.append(temp)
                return

            temp+='('
            solve(temp, cntx-1, cnty)
            temp = temp[:-1]

            temp+=')'
            solve(temp, cntx, cnty-1)
            temp = temp[:-1]
        
        solve("", n, n)

        return ans

        """
        :type n: int
        :rtype: List[str]
        """
        

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public void solve(List<String> ans, StringBuilder temp, int lc, int rc){
        if(lc<0 || rc<0) return;
        if(lc==0 && rc==0){
            ans.add(temp.toString());
            return;
        }
        if(lc < rc){
            temp.append('(');
            solve(ans,temp, lc-1, rc);
            temp.deleteCharAt(temp.length()-1);
            temp.append(')');
            solve(ans,temp, lc, rc-1);
            temp.deleteCharAt(temp.length()-1);
        }else{
            temp.append('(');
            solve(ans,temp, lc-1, rc);
            temp.deleteCharAt(temp.length()-1);
        }
        return;
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<String>();
        StringBuilder temp = new StringBuilder();
        solve(ans, temp, n , n);
        return ans;
    }
}